das/checkpoint_store.go

```diff
@@ -13,35 +13,44 @@ var (
 	checkpointKey = datastore.NewKey("checkpoint")
 )
 
+type checkpoint struct {
+	MinSampled uint64         `json:"min_sampled_height"` // lowest sampled height
+	MaxKnown   uint64         `json:"max_known_height"`   // height of the newest known header
+	Skipped    map[uint64]int `json:"skipped"`            // header's heights that been skipped with corresponding try count
+}
+
 // wrapCheckpointStore wraps the given datastore.Datastore with the `das`
 // prefix. The checkpoint store stores/loads the DASer's checkpoint to/from
-// disk using the checkpointKey. The checkpoint is stored as a uint64
+// disk using the checkpointKey. The checkpoint is stored as an uint64
 // representation of the height of the latest successfully DASed header.
 func wrapCheckpointStore(ds datastore.Datastore) datastore.Datastore {
 	return namespace.Wrap(ds, storePrefix)
 }
 
-// loadCheckpoint loads the DAS checkpoint height from disk and returns it.
+// loadCheckpoint loads the DAS checkpoint from disk and returns it.
 // If there is no known checkpoint, it returns height 0.
-func loadCheckpoint(ctx context.Context, ds datastore.Datastore) (int64, error) {
-	checkpoint, err := ds.Get(ctx, checkpointKey)
+func loadCheckpoint(ctx context.Context, ds datastore.Datastore) (checkpoint, error) {
+	bs, err := ds.Get(ctx, checkpointKey)
 	if err != nil {
-		// if no checkpoint was found, return checkpoint as
-		// 0 since DASer begins sampling on checkpoint+1
+		// if no checkpoint was found, return zero-value checkpoint
 		if err == datastore.ErrNotFound {
 			log.Debug("checkpoint not found, starting sampling at block height 1")
-			return 0, nil
+			return checkpoint{}, nil
 		}
 
-		return 0, err
+		return checkpoint{}, err
 	}
-	return int64(binary.BigEndian.Uint64(checkpoint)), err
+
+	cp := checkpoint{}
+	err = json.Unmarshal(bs, &cp)
+	return cp, err
 }
 
```

Comment:
If we always return zero value here, we might return `json.Unmarshal(bs, &cp)` diretly.